// @rei/cedar/dist/lib/components/rating/CdrRating.mjs
import clsx from 'clsx';
import _concatInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/concat';
import _mapInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/map';
import size from '../../mixins/size.mjs';
import space from '../../mixins/space.mjs';
import contentPriority from '../../mixins/contentPriority.mjs';
import _keysInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/keys';
import _toConsumableArray from '@babel/runtime-corejs3/helpers/esm/toConsumableArray';
import style from './styles/CdrRating.mjs';

var CdrRating = {
  name: 'CdrRating',
  mixins: [contentPriority, size, space],
  props: {
    /**
     * Rating value (out of 5)
     */
    rating: {
      required: true,
      type: [String, Number],
      default: 0
    },

    /**
     * Total number of ratings
     */
    count: {
      required: false,
      type: [String, Number],
      default: null
    },

    /**
     * Hides the word 'reviews' if true
     */
    compact: {
      type: Boolean,
      default: false
    },

    /**
     * Allows the ratings to act as a link
     */
    href: {
      type: String
    }
  },
  data: function data() {
    return {
      style: style
    };
  },
  computed: {
    baseClass: function baseClass() {
      return 'cdr-rating';
    },
    rounded: function rounded() {
      return Math.round(this.rating * 4) / 4;
    },
    whole: function whole() {
      return Math.floor(this.rounded);
    },
    remainder: function remainder() {
      return this.rounded.toFixed(2).split('.')[1];
    },
    empties: function empties() {
      return 5 - this.whole - (this.remainder > 0 ? 1 : 0);
    },
    formattedCount: function formattedCount() {
      return this.compact ? "(".concat(this.count, ")") : "".concat(this.count);
    },
    remainderEl: function remainderEl() {
      var h = this.$createElement;
      var remainder;

      if (this.remainder === '25') {
        remainder = h("span", {
          "class": clsx(this.style['cdr-rating__icon'], this.style['cdr-rating__25']),
          "attrs": {
            "aria-hidden": "true"
          }
        });
      } else if (this.remainder === '50') {
        remainder = h("span", {
          "class": clsx(this.style['cdr-rating__icon'], this.style['cdr-rating__50']),
          "attrs": {
            "aria-hidden": "true"
          }
        });
      } else if (this.remainder === '75') {
        remainder = h("span", {
          "class": clsx(this.style['cdr-rating__icon'], this.style['cdr-rating__75']),
          "attrs": {
            "aria-hidden": "true"
          }
        });
      }

      return remainder;
    },
    ratingSrText: function ratingSrText() {
      return "rated ".concat(this.rounded, " out of 5");
    },
    countSrText: function countSrText() {
      return this.count !== null ? " with ".concat(this.count, " reviews") : '';
    }
  },
  render: function render() {
    var _context,
        _context2,
        _this = this,
        _context4,
        _context5;

    var h = arguments[0];
    var Component = this.href ? 'a' : 'div';
    return h(Component, {
      "attrs": {
        "href": this.href
      },
      "class": clsx(this.style[this.baseClass], this.contentPriorityClass, this.sizeClass, this.space, this.href ? this.style['cdr-rating--linked'] : '')
    }, [h("div", {
      "class": this.style['cdr-rating__ratings']
    }, [_mapInstanceProperty(_context = _toConsumableArray(_keysInstanceProperty(_context2 = Array(this.whole)).call(_context2))).call(_context, function (n) {
      var _context3;

      return h("span", {
        "class": clsx(_this.style['cdr-rating__icon'], _this.style['cdr-rating__100']),
        "key": _concatInstanceProperty(_context3 = "rating-whole-".concat(n, "-")).call(_context3, _this._uid),
        "attrs": {
          "aria-hidden": "true"
        }
      });
    }), this.remainderEl, _mapInstanceProperty(_context4 = _toConsumableArray(_keysInstanceProperty(_context5 = Array(this.empties)).call(_context5))).call(_context4, function (n) {
      var _context6;

      return h("span", {
        "class": clsx(_this.style['cdr-rating__icon'], _this.rounded > 0 || _this.count > 0 ? _this.style['cdr-rating__placeholder'] : _this.style['cdr-rating__placeholder--no-reviews']),
        "key": _concatInstanceProperty(_context6 = "rating-empty-".concat(n, "-")).call(_context6, _this._uid),
        "attrs": {
          "aria-hidden": "true"
        }
      });
    })]), this.count !== null ? h("span", {
      "attrs": {
        "aria-hidden": "true"
      },
      "class": this.style['cdr-rating__count']
    }, [this.href && h("span", {
      "class": this.style['cdr-rating__number']
    }, [this.rounded]), h("span", [this.formattedCount]), !this.compact && h("span", ["\xA0Reviews"])]) : '', h("span", {
      "class": this.style['cdr-rating__caption-sr']
    }, [this.ratingSrText, this.countSrText])]);
  }
};

export default CdrRating;
