// @rei/cedar/dist/lib/components/list/CdrList.mjs
import clsx from 'clsx';
import space from '../../mixins/space.mjs';
import modifier from '../../mixins/modifier.mjs';
import _indexOfInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/index-of';
import contentPriority from '../../mixins/contentPriority.mjs';
import style from './styles/CdrList.mjs';

var CdrList = {
  name: 'CdrList',
  mixins: [modifier, space, contentPriority],
  props: {
    /**
     * Tag accepts a user defined element and expects either: 'ul' = Unordered List or 'ol' = Ordered List.
     */
    tag: {
      type: String,
      default: 'ul',
      validator: function validator(value) {
        var _context;

        return _indexOfInstanceProperty(_context = ['ul', 'ol']).call(_context, value) >= 0 || false;
      }
    }
  },
  data: function data() {
    return {
      style: style
    };
  },
  computed: {
    baseClass: function baseClass() {
      return 'cdr-list';
    }
  },
  render: function render() {
    var h = arguments[0];
    var Component = this.tag;
    return h(Component, {
      "class": clsx(this.style[this.baseClass], this.modifierClass, this.space, this.contentPriorityClass)
    }, [this.$slots.default]);
  }
};

export default CdrList;
