// @rei/cedar/dist/lib/components/cta/CdrCta.mjs
import clsx from 'clsx';
import space from '../../mixins/space.mjs';
import modifier from '../../mixins/modifier.mjs';
import _indexOfInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/index-of';
import fullWidth from '../../mixins/fullWidth.mjs';
import style from './styles/CdrCta.mjs';

var CdrCta = {
  name: 'CdrCta',
  mixins: [modifier, space, fullWidth],
  props: {
    /**
      * Change the color of the cdr-cta button match different themes.
      */
    ctaStyle: {
      type: String,
      default: 'dark',
      validator: function validator(value) {
        var _context;

        return _indexOfInstanceProperty(_context = ['brand', 'dark', 'light', 'sale']).call(_context, value) >= 0 || false;
      }
    },
    href: {
      type: String,
      default: '#'
    },

    /** @ignore */
    target: String,

    /** @ignore */
    rel: String
  },
  data: function data() {
    return {
      style: style
    };
  },
  computed: {
    baseClass: function baseClass() {
      return 'cdr-cta';
    },
    ctaClass: function ctaClass() {
      return this.modifyClassName(this.baseClass, this.ctaStyle);
    },
    computedRel: function computedRel() {
      if (this.target === '_blank') {
        return this.rel || 'noopener noreferrer';
      }

      return this.rel;
    }
  },
  render: function render() {
    var h = arguments[0];
    return h("a", {
      "class": clsx(this.style[this.baseClass], this.modifierClass, this.ctaClass, this.fullWidthClass, this.space),
      "attrs": {
        "target": this.target,
        "rel": this.computedRel,
        "href": this.href
      }
    }, [this.$slots.default, h("svg", {
      "attrs": {
        "xmlns": "http://www.w3.org/2000/svg",
        "viewBox": "0 0 24 24",
        "role": "presentation"
      },
      "class": this.style['cdr-cta__icon']
    }, [h("path", {
      "attrs": {
        "d": "M16 12a.997.997 0 0 0-.288-.702l-5.005-5.005a1 1 0 0 0-1.414 1.414L13.585 12 9.29 16.295a1 1 0 0 0 1.417 1.412l4.98-4.98A.997.997 0 0 0 16 12z"
      }
    })])]);
  }
};

export default CdrCta;
